src/libraries/System.Net.WebSockets.Client/tests/DeflateTests.cs

```diff
@@ -86,6 +86,51 @@ public DeflateTests(ITestOutputHelper output) : base(output)
             }), new LoopbackServer.Options { WebSocketEndpoint = true });
         }
 
+        [ConditionalFact(nameof(WebSocketsSupported))]
+        public async Task ThrowsWhenContinuationHasDifferentCompressionFlags()
+        {
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                using var cws = new ClientWebSocket();
+                using var cts = new CancellationTokenSource(TimeOutMilliseconds);
+
+                cws.Options.DangerousDeflateOptions = new WebSocketDeflateOptions();
+                await ConnectAsync(cws, uri, cts.Token);
+
+                await cws.SendAsync(Memory<byte>.Empty, WebSocketMessageType.Text, WebSocketMessageFlags.DisableCompression, default);
+                Assert.Throws<ArgumentException>("messageFlags", () =>
+                   cws.SendAsync(Memory<byte>.Empty, WebSocketMessageType.Binary, WebSocketMessageFlags.EndOfMessage, default));
+            }, server => server.AcceptConnectionAsync(async connection =>
+            {
+                await LoopbackHelper.WebSocketHandshakeAsync(connection);
+            }), new LoopbackServer.Options { WebSocketEndpoint = true });
+        }
+
+        [ConditionalFact(nameof(WebSocketsSupported))]
+        public async Task SendHelloWithDisableCompression()
+        {
+            byte[] bytes = "Hello"u8.ToArray();
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                using var cws = new ClientWebSocket();
+                using var cts = new CancellationTokenSource(TimeOutMilliseconds);
+
+                cws.Options.DangerousDeflateOptions = new WebSocketDeflateOptions();
+                await ConnectAsync(cws, uri, cts.Token);
+
+                WebSocketMessageFlags flags = WebSocketMessageFlags.DisableCompression | WebSocketMessageFlags.EndOfMessage;
+                await cws.SendAsync(bytes, WebSocketMessageType.Text, flags, cts.Token);
+            }, server => server.AcceptConnectionAsync(async connection =>
+            {
+                var buffer = new byte[bytes.Length];
+                await LoopbackHelper.WebSocketHandshakeAsync(connection);
```

Comment:
This test doesn't seem to negotiate compression from server side... so this whole ws connection is uncompressed, and the flag doesn't do anything. You need to include extensions in the reply as in the first test in this file https://github.com/dotnet/runtime/blob/8fe12bc5a66a5ef42200a3d88d68b5991eb158e6/src/libraries/System.Net.WebSockets.Client/tests/DeflateTests.cs#L82

I also believe we should check in the same test, that a message without the flag would be properly compressed, and only the flagged one would be uncompressed.

Reply:
It is uncompressed or client will send compressed message after all? I found in another test that client send compressed and server should throw https://github.com/dotnet/runtime/blob/0c4ee9e749167a2185f59b4a16c8a5c62e0ec5df/src/libraries/System.Net.WebSockets/tests/WebSocketDeflateTests.cs#L388-L390

Reply:
The test you're linking is with core `WebSocket` class -- it doesn't handle the handshake and just "believes" whatever is passed to the constructor. `ClientWebSocket` handles the handshake including extensions negotiation -- if server never replies with negotiated compression, it would mean that it doesn't support it, so `negotiatedDeflateOptions` would never be created and applied to the upderlying core `WebSocket` instance, see
https://github.com/dotnet/runtime/blob/b90a734e1c7d7d1e0a7e1a25ca054a3b76dd2dfa/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L176-L189